tests/python_package_test/test_basic.py

```diff
@@ -89,6 +91,115 @@ def test_basic(tmp_path):
                                    bst.predict, tname)
 
 
+class NumpySequence(lgb.Sequence):
+    def __init__(self, ndarray, batch_size):
+        self.ndarray = ndarray
+        self.batch_size = batch_size
+
+    def __getitem__(self, idx):
+        # The simple implementation is just a single "return self.ndarray[idx]"
+        # The following is for demo and testing purpose.
+        if isinstance(idx, numbers.Integral):
+            return self._get_one_line(idx)
+        elif isinstance(idx, slice):
+            if not (idx.step is None or idx.step is 1):
+                raise NotImplementedError("No need to implement, caller will not set step by now")
+            return self.ndarray[idx.start: idx.stop]
+        else:
+            raise TypeError(f"Sequence Index must be an integer/list/slice, got {type(idx)}")
+
+    def _get_one_line(self, idx):
+        return self.ndarray[idx]
+
+    def __len__(self):
+        return len(self.ndarray)
+
+
+def _create_sequence_from_ndarray(data, num_seq, batch_size):
+    if num_seq == 1:
+        return NumpySequence(data, batch_size)
+
+    nrow = data.shape[0]
+    seqs = []
+    seq_size = nrow // num_seq
+    for start in range(0, nrow, seq_size):
+        end = min(start + seq_size, nrow)
+        seq = NumpySequence(data[start:end], batch_size)
+        seqs.append(seq)
+    return seqs
+
+
+@pytest.mark.parametrize('sample_count', [11, 23, 100, None])
+@pytest.mark.parametrize('batch_size', [3, 20, None])
+@pytest.mark.parametrize('include_0', [False, True])
+@pytest.mark.parametrize('include_nan', [False, True])
+@pytest.mark.parametrize('num_seq', [1, 3])
+def test_sequence(tmpdir, sample_count, batch_size, include_0, include_nan, num_seq):
+    params = {'bin_construct_sample_cnt': sample_count}
+
+    nrow = 50
+    half_nrow = nrow // 2
+    ncol = 11
+    data = np.arange(nrow * ncol, dtype=np.float64).reshape((nrow, ncol))
+
+    # total col
+    if include_0:
+        data[:, 0] = 0
+    if include_nan:
+        data[:, 1] = np.nan
+
+    # half col
+    if include_nan:
+        # nan col
+        data[0:half_nrow, 2] = np.nan
+    if include_0:
+        # 0 col
+        data[0:half_nrow, 3] = 0
+
+    # nan + 0 col
+    if include_nan:
+        data[0:half_nrow, 4] = np.nan
+    if include_0:
+        data[half_nrow:-2, 4] = 0
+
+    X = data[:, :-1]
+    Y = data[:, -1]
+
+    npy_bin_fname = os.path.join(tmpdir, 'data_from_npy.bin')
+    seq_bin_fname = os.path.join(tmpdir, 'data_from_seq.bin')
+
+    # Create dataset from numpy array directly.
+    ds = lgb.Dataset(X, label=Y, params=params)
+    ds.save_binary(npy_bin_fname)
+
+    # Create dataset using Sequence.
+    seqs = _create_sequence_from_ndarray(X, num_seq, batch_size)
+    seq_ds = lgb.Dataset(seqs, label=Y, params=params)
+    seq_ds.save_binary(seq_bin_fname)
+
+    assert filecmp.cmp(npy_bin_fname, seq_bin_fname)
+
+    # Test for validation set.
```

Comment:
Sorry, can get it. Is validation dataset creation supported (#4184)?
If so, please add `Sequence` here 
https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/python-package/lightgbm/basic.py#L1477

Reply:
Yes. We can create valid dataset from `Sequence`. I'll update doc.

Reply:
Was #4184 created for discussion purposes or that issue is for some other feature request? In other words, can we close it after merging this PR?

Reply:
It's more for ciscussion purpose.